### PR TITLE
bpo-4260: Document that ctypes.xFUNCTYPE are decorators

### PR DESCRIPTION
**Pull request by andresdelfino**
https://bugs.python.org/issue4260

### CODE REVIEW

File: Doc/library/ctypes.rst [hunk 1]
```diff
@@ -1582,7 +1599,7 @@ type and the argument types of the function.
 
 .. function:: CFUNCTYPE(restype, *argtypes, use_errno=False, use_last_error=False)
 
-   The returned function prototype creates functions that use the standard C
+   Returns a function prototype which creates functions that use the standard C
```
Comment (berkerpeksag) [inline Doc/library/ctypes.rst]: Please don't make changes irrelevant to bpo-4260.

File: Doc/library/ctypes.rst [hunk 2]
```diff
@@ -1592,7 +1609,7 @@ type and the argument types of the function.
 
 .. function:: WINFUNCTYPE(restype, *argtypes, use_errno=False, use_last_error=False)
 
-   Windows only: The returned function prototype creates functions that use the
+   Windows only: Returns a function prototype which creates functions that use the
```
Comment (berkerpeksag) [inline Doc/library/ctypes.rst]: Please don't make changes irrelevant to bpo-4260.

File: Doc/library/ctypes.rst [hunk 3]
```diff
@@ -1601,9 +1618,12 @@ type and the argument types of the function.
 
 .. function:: PYFUNCTYPE(restype, *argtypes)
 
-   The returned function prototype creates functions that use the Python calling
+   Returns a function prototype which creates functions that use the Python calling
```
Comment (berkerpeksag) [inline Doc/library/ctypes.rst]: Please don't make changes irrelevant to bpo-4260.

File: Doc/library/ctypes.rst [hunk 4]
```diff
@@ -1601,9 +1618,12 @@ type and the argument types of the function.
...
    convention.  The function will *not* release the GIL during the call.
 
+These factory functions can be used as decorator factories, and as such, be applied
```
Comment (berkerpeksag) [inline Doc/library/ctypes.rst]: I think we can move this information to the first paragraph of the "Function prototypes" section to make it more discoverable.

Also, we can add a link to the examples from that paragraph.

File: Doc/library/ctypes.rst [hunk 5]
```diff
@@ -1025,6 +1025,23 @@ As we can easily check, our array is sorted now::
    1 5 7 33 99
    >>>
 
+The function factories can be used as decorator factories, so we may as well
+write::
+
+   >>> @CFUNCTYPE(c_int, POINTER(c_int), POINTER(c_int))
+   ... def py_cmp_func(*args):
```
Comment (berkerpeksag) [inline Doc/library/ctypes.rst]: Why didn't you reuse the `py_cmp_func` function in line 1008? IMO, the following snippet should work:

```py
@CFUNCTYPE(c_int, POINTER(c_int), POINTER(c_int))
def py_cmp_func(a, b):
    print("py_cmp_func", a[0], b[0])
    return a[0] - b[0]
```

There is even a test for that same function:

https://github.com/python/cpython/blob/b91a3a0d61596cafb1b46c98fab65fee16a8bbbb/Lib/ctypes/test/test_callbacks.py#L186-L188

**Review by berkerpeksag — CHANGES_REQUESTED** [review]

File: Doc/library/ctypes.rst (hunk 5 above)
Comment (andresdelfino) [reply Doc/library/ctypes.rst]: I followed the proposal from David W. Lambert in the issue, but you are right, I tested the decorator with the function from line 1008 and works as well.

**Comment by andresdelfino** [comment]
I have made the requested changes; please review again 🐍🐍🐍

File: Doc/library/ctypes.rst [hunk 6]
```diff
@@ -1577,7 +1593,10 @@ Foreign functions can also be created by instantiating function prototypes.
 Function prototypes are similar to function prototypes in C; they describe a
 function (return type, argument types, calling convention) without defining an
 implementation.  The factory functions must be called with the desired result
-type and the argument types of the function.
+type and the argument types of the function, and can be used as decorator
+factories, and as such, be applied to functions through the ``@wrapper`` syntax.
+See :ref:`ctypes-callback-functions` for examples.
+
```
Comment (berkerpeksag) [inline Doc/library/ctypes.rst]: Style nit: We can drop this extra newline.

**Review by berkerpeksag — APPROVED** [review]

**Comment by berkerpeksag** [comment]
This looks pretty good to me, thank you! I will wait for a day or two to let others comment on the PR. Please ping me again if I don't merge this by Friday.

**Comment by berkerpeksag** [comment]
@andresdelfino thanks!

**Pull request merged**